runner/operator/argos_operator/v2_2_0/bin/retrieve_samples_by_query.py

```diff
@@ -0,0 +1,351 @@
+    if descriptor:
+        pooled_normals = FileRepository.filter(
+            queryset=pooled_normals, metadata={settings.RECIPE_METADATA_KEY: descriptor}
+        )
+
+        # sample_name is FROZENPOOLEDNORMAL unless FFPE is in any of the preservation types
+        # in preservation_types
+        preservations_lower_case = set([x.lower() for x in preservation_types])
+        run_ids_suffix_list = [i for i in run_ids if i]  # remove empty or false string values
+        run_ids_suffix = "_".join(set(run_ids_suffix_list))
+        sample_name = "FROZENPOOLEDNORMAL_" + run_ids_suffix
+        if "ffpe" in preservations_lower_case:
+            sample_name = "FFPEPOOLEDNORMAL_" + run_ids_suffix
+    else:
+        machine = get_pooled_normal_machine(run_ids)
+        preservations_lower_case = set([x.lower() for x in preservation_types])
+        preservation = "frozen"
+        baits = bait_set.lower()
+        if "ffpe" in preservations_lower_case:
+            preservation = "ffpe"
+        # This will retrieve one entry in the file_system_poolednormal table
+        pooled_normal_objs = PooledNormal.objects.filter(
+            machine=machine, preservation_type=preservation, bait_set=baits
+        )
+        pooled_normals = list()
+        all_files = FileRepository.filter(file_group=settings.POOLED_NORMAL_FILE_GROUP)
+        for obj in pooled_normal_objs:
+            for pn in obj.pooled_normals_paths:
+                pn_file = FileRepository.filter(queryset=all_files, path=pn).first()
+                pooled_normals.append(pn_file)
+        if machine:
+            sample_name = "_".join([bait_set, preservation, machine, "POOLEDNORMAL"])
+            sample_name = sample_name.upper()
+            descriptor = baits
+
+    return pooled_normals, descriptor, sample_name
+
+
+def get_sequencer_type(run_ids_list):
+    run_ids_lower = [i.lower() for i in run_ids_list if i]
+    machine_modes = MachineRunMode.objects.all()
+    for machine in machine_modes:
+        if find_substr(machine.machine_name, run_ids_lower):
+            return machine.machine_class
+
+
+def get_pooled_normal_machine(run_ids_list):
+    run_ids_lower = [i.lower() for i in run_ids_list if i]
+    pooled_normals = PooledNormal.objects.all()
+    earliest_pooled_normal = None
+    for pooled_normal in pooled_normals:
+        if find_substr(pooled_normal.machine, run_ids_lower):
+            if not earliest_pooled_normal:
+                earliest_pooled_normal = pooled_normal
+            else:
+                # tiebreaker - if there's more than one set of pooled normals fetched, return earliest one
+                old_date = earliest_pooled_normal.run_date
+                curr_date = pooled_normal.run_date
+                if curr_date < old_date:
+                    earliest_pooled_normal = pooled_normal
+    if earliest_pooled_normal:
+        machine = earliest_pooled_normal.machine
+        return machine
+    return None
```

Comment:
This retrieves the pooled normal from the PooledNormal model. It also performs the tiebreaker - see comment.

Reply:
This was actually bothering me so I refactored it. Will resolve this and add the new logic here.